### PR TITLE
fuse: handle errors appropriately when getting group list

### PR DESCRIPTION
**Pull request by jtlayton**
Like this PR, but more along the lines of what @ukernel suggested:

https://github.com/ceph/ceph/pull/21025

If fetching groups fails, just log to the debug log and carry on.

--------------------8<---------------------

Currently, the GET_GROUPS macro just ignores errors that occur when
fetching the current task's groups. The callers then end up casting
the error to a count.

Change the macro to a real function, and just have it cleanly ignore
errors when getting groups fails. We do log the error to the debug
log however in that case.

Tracker: http://tracker.ceph.com/issues/23446
Signed-off-by: Jeff Layton <jlayton@redhat.com>

### CODE REVIEW

File: src/client/fuse_ll.cc [hunk 1]
```diff
@@ -136,13 +136,21 @@ static int getgroups_cb(void *handle, gid_t **sgids)
...
 }
 
-#define GET_GROUPS(perms, req)	{				\
-  if (g_conf->get_val<bool>("fuse_set_user_groups")) {	\
-    gid_t *gids = NULL;						\
-    int count = getgroups(req, &gids);				\
-    perms.init_gids(gids, count);				\
-    perms.take_gids();						\
-  } }
+static int get_fuse_groups(UserPerm& perms, fuse_req_t req)
```
Comment (joscollin) [inline src/client/fuse_ll.cc]: returns `int` ?

Comment (joscollin) [reply src/client/fuse_ll.cc]: Sorry, I got it. :-)

Comment (jtlayton) [reply src/client/fuse_ll.cc]: No, you're right. That should be void return. Will fix.

**Comment by jtlayton** [comment]
Retest this please.

**Comment by joscollin** [comment]
retest this please

File: src/client/fuse_ll.cc (hunk 1 above)
Comment (joscollin) [reply src/client/fuse_ll.cc]: @jtlayton It was not there in the previous jenkins build. I'm not sure why. 

But I got it in my local build:
```
$ make client/fuse_ll.cc.o
/home/jcollin/workspace/ceph/src/client/fuse_ll.cc: In function ‘int get_fuse_groups(UserPerm&, fuse_req_t)’:
/home/jcollin/workspace/ceph/src/client/fuse_ll.cc:153:1: warning: no return statement in function returning non-void [-Wreturn-type]
 }
 ^
```

**Comment by jtlayton** [comment]
Add a couple of more commits to clean up UserPerm code a bit.

**Comment by jtlayton** [comment]
Not sure what's up with the test failures on this. It looks like maybe the mgr dashboard build is broken? This PR is based on a pull from a couple of weeks ago though, so I imagine that that should be fixed by now. I can rebase this if you like, but github PRs don't seem to like that.

**Comment by joscollin** [comment]
@jtlayton 

make check failure of PRs due to pip
https://github.com/ceph/ceph/pull/21159

**Comment by jtlayton** [comment]
Squashed down the patches, so this should be ready for merge.

**Comment by renhwztetecs** [comment]
looks good,  I encountered the same problem when using the ceph-fuse mount on the Alpine system.

**Comment by ukernel** [comment]
LTGM

**Comment by jtlayton** [comment]
One more patch. Remove the unused getgroups_cb.

**Comment by jtlayton** [comment]
Minor update to last patch -- remove client_getgroups_callback_t definition.

**Pull request merged**